// homebridge-gsh/dist/types/humidity-sensor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HumiditySensor = void 0;
const hap_types_1 = require("../hap-types");
class HumiditySensor {
    sync(service) {
        return {
            id: service.uniqueId,
            type: 'action.devices.types.SENSOR',
            traits: [
                'action.devices.traits.HumiditySetting',
            ],
            name: {
                defaultNames: [
                    service.serviceName,
                    service.accessoryInformation.Name,
                ],
                name: service.serviceName,
                nicknames: [],
            },
            willReportState: true,
            attributes: {
                queryOnlyHumiditySetting: true,
            },
            deviceInfo: {
                manufacturer: service.accessoryInformation.Manufacturer,
                model: service.accessoryInformation.Model,
                hwVersion: service.accessoryInformation.HardwareRevision,
                swVersion: service.accessoryInformation.SoftwareRevision,
            },
            customData: {
                aid: service.aid,
                iid: service.iid,
                instanceUsername: service.instance.username,
                instanceIpAddress: service.instance.ipAddress,
                instancePort: service.instance.port,
            },
        };
    }
    query(service) {
        var _a;
        return {
            online: true,
            humidityAmbientPercent: (_a = service.characteristics.find(x => x.type === hap_types_1.Characteristic.CurrentRelativeHumidity)) === null || _a === void 0 ? void 0 : _a.value,
        };
    }
    execute(service, command) {
        return { payload: { characteristics: [] } };
    }
}
exports.HumiditySensor = HumiditySensor;
